### PR TITLE
add oVirt image for CI tests

### PR DESCRIPTION
**Pull request by michalskrivanek**
can you please take a look at the approach? 
The setup script is a bit convoluted, but seems to be working reliably and the simple stub check-ovirt seem to bring everything up to a point where ovirt-machines can do its job
image-create has 2GB RAM which is not really much for ovirt-engine, even with 4GV the creating takes wuite some time
@mareklibra can pick up from here to write some useful tests

### CODE REVIEW

**Comment by michalskrivanek** [comment]
few questions - would anyone mind bumping up the memory size of the VM used for image-create from 2GB to 4GB?  It starts the whole thing and 4GB is really the minimum...it does start up with 2GB, but it may have stability issues and it's definitely slower. 
Another option would be to move all the initial setup into image-prepare...but...then image-prepare is going to take way longer (regardless the memory size it's quite slow)

**Comment by michalskrivanek** [comment]
also - would anyone mind adding virtio-rng device to the Machine class VM definition? It shouldn't hurt anybody, and it speeds up crypto operations (we create some certificates as part of ovirt-engine deployment and it hangs for quite some time due to lack of entropy)

**Comment by stefwalter** [comment]
> also - would anyone mind adding virtio-rng device to the Machine class VM definition? It shouldn't hurt anybody, and it speeds up crypto operations (we create some certificates as part of ovirt-engine deployment and it hangs for quite some time due to lack of entropy)

Not against it. I guess you've found the place in testvm.py where it would be done?

**Comment by michalskrivanek** [comment]
pushed #7480 for virtio-rng change

File: test/verify/check-ovirt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# DC: "Default"
+# Cluster: "Default"
+# Host: "host" running in the image with plaintext communication to engine
+# VM: "VM" Blank-based VM with network boot 
+#
+
+WAIT_SCRIPT = """
+set -ex
+for x in $(seq 1 60); do
+    if [ "$(curl -s -k -u admin@internal:engine https://%(addr)s/ovirt-engine/api | xmllint --xpath "string(/api/product_info/name)" -)" = "oVirt Engine" ]; then
```
Comment (martinpitt) [inline test/verify/check-ovirt]: xmllint smells like a new test dependency. Could this be replaced with a simple grep?

If that's not practical, can you please add it to HACKING? We would also need to update the cockpit test container for that, so I'd like to avoid it.

File: test/verify/check-ovirt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        MachineCase.setUp(self)
+        self.ovirt = self.machines['ovirt']
+        m = self.machine
+
+        # Wait for the web service to be accessible
+        args = { "addr": "10.111.113.5" }
+        m.execute(script=WAIT_SCRIPT % args)
+
+        vmid = m.execute("curl -s -k -u admin@internal:engine -H Content-type:application/xml https://10.111.113.5/ovirt-engine/api/vms/ | xmllint --xpath \"string(/vms/vm[name=\'VM\']/@id)\" -")
+	self.assertNotEqual(vmid, "")
```
Comment (martinpitt) [inline test/verify/check-ovirt]: apparently there's a tab here (whitespace glitch)

File: test/verify/check-ovirt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        m.execute("curl -s -k -u admin@internal:engine -H Content-type:application/xml -X POST -d \"<action/>\"  https://10.111.113.5/ovirt-engine/api/vms/" + vmid + "/start")
+
+    def testBasic(self):
+        b = self.browser
+
+        self.login_and_go("/machines")
```
Comment (martinpitt) [inline test/verify/check-ovirt]: Heh, yes, that needs some filig out :-) But you can add a `sit()` here and do some manual testing with `--sit`; does that work, i. e. does it bring up and connect to the ovirt image properly?

File: bots/images/scripts/ovirt.install [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#! /bin/bash
+
+set -e
+
+/var/lib/testvm/fedora.install --rhel "$@"
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.install]: I don't think you want `--rhel` here, you are starting from a centos image.

File: bots/images/scripts/ovirt.setup [hunk 5]
```diff
@@ -0,0 +1,151 @@
+#!/bin/bash
+
+set -eux
+
+echo foobar | passwd --stdin root
+
+/usr/bin/systemctl stop firewalld NetworkManager kdump
+/usr/bin/systemctl disable firewalld NetworkManager kdump
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: Please avoid hardcoded paths; This very plausibly will move to `/bin/` in newer OSes. There are plenty more instances below, I don't mention them all individually.

File: bots/images/scripts/ovirt.setup [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+iptables -F
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: This is supposed to disable the firewall during image installation? Why is that necessary?

File: bots/images/scripts/ovirt.setup [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+FQDN=ovirt.cockpit.lan
+hostnamectl set-hostname ${FQDN}
+
+# Latest stable repos
+rpm -ivh http://resources.ovirt.org/pub/yum-repo/ovirt-release41.rpm
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: Hardcoding the release number should be avoided; can there be some code which scans available releases and takes the latest one?

File: bots/images/scripts/ovirt.setup [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+rpm --import /etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: How is the CentOS-7 key not already known to rpm on that image?

File: bots/images/scripts/ovirt.setup [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+rpm --import /etc/pki/rpm-gpg/RPM-GPG-ovirt-*
+rpm --import https://dl.fedoraproject.org/pub/epel/RPM-GPG-KEY-EPEL-7
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: Likewise, this looks too low-level; shouldn't that be done by something like `yum-config-manager --enable epel..`?

File: bots/images/scripts/ovirt.setup [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+OVESETUP_CONFIG/engineServiceStopNeeded=bool:False
+OVESETUP_ASYNC/clearTasks=bool:False
+OVESETUP_DWH_CORE/enable=bool:False
+OVESETUP_DWH_PROVISIONING/postgresProvisioningEnabled=bool:False
+OVESETUP_CONFIG/imageioProxyConfig=bool:False
+# Don't install the the OVN provider:
+OVESETUP_OVN/ovirtProviderOvn=bool:False
+# Don't perform database full vacuum:
+OVESETUP_DB/engineVacuumFull=bool:False
+EOF
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: That's a lot of customization. Does it somehow work to use the internal defaults for most settings and just set the actually important parts here?

File: bots/images/scripts/ovirt.setup [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+rpm --import https://raw.githubusercontent.com/CentOS-Storage-SIG/centos-release-storage-common/master/RPM-GPG-KEY-CentOS-SIG-Storage
+yum install -y vdsm vdsm-client vdsm-hook-faqemu
+cat << 'EOF' > /etc/vdsm/vdsm.conf
+[vars]
+ssl = false
+fake_kvm_support = true
+EOF
+# WA until commit bf0f31e2b
+sed -i "s/pc-i440fx-rhel7.2.0/pc-i440fx-rhel7.3.0/g" /usr/libexec/vdsm/hooks/after_get_caps/10_faqemu
+sed -i "/def _graphics_spice_to_vnc(/i @workaround()" /usr/libexec/vdsm/hooks/before_vm_start/10_faqemu
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: Changing files in `/usr`? That seems wrong, these are under the purview of rpm packages.

File: bots/images/scripts/ovirt.setup [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+# add host, create storage on /nfsexport, attach it
+CURL="/usr/bin/curl -k -u admin@internal:engine -H Content-type:application/xml"
+DCID=$(${CURL} "https://${FQDN}/ovirt-engine/api/datacenters?search=name%3DDefault" | xmllint --xpath "string(/data_centers/data_center/@id)" -)
+${CURL} -d "<host> <name>host</name> <address>${FQDN}</address> <root_password>foobar</root_password> </host>" https://${FQDN}/ovirt-engine/api/hosts
+HOST_UP=""
+while [ -z ${HOST_UP} ]; do
+    sleep 5
+    HOST_UP=$(${CURL} "https://ovirt.cockpit.lan/ovirt-engine/api/hosts?search=status%3Dup" | xmllint --xpath "string(/hosts/host/@id)" -)
+done
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: Potentially infinite loop. Please rather use something like `for retry in $(seq 60)` or so, to avoid forever-hanging image rebuilds if errors occur.

File: bots/images/scripts/ovirt.setup [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+STORAGEID=$(${CURL} -d "<storage_domain> <name>data</name> <type>data</type> <storage> <type>nfs</type> <address>${FQDN}</address> <path>/nfsexport</path> </storage> <host> <name>host</name> </host> </storage_domain>" https://${FQDN}/ovirt-engine/api/storagedomains | xmllint --xpath "string(/storage_domain/@id)" -)
+${CURL} -d "<storage_domain> <name>data</name> </storage_domain>" https://${FQDN}/ovirt-engine/api/datacenters/${DCID}/storagedomains
+
+# wait on DC to go up, may take a while
+DC_UP=""
+while [ -z ${DC_UP} ]; do
+    sleep 5
+    DC_UP=$(${CURL} "https://ovirt.cockpit.lan/ovirt-engine/api/datacenters?search=status%3Dup" | xmllint --xpath "string(/data_centers/data_center/@id)" -)
+done
```
Comment (martinpitt) [inline bots/images/scripts/ovirt.setup]: Same issue.

File: test/verify/check-ovirt (hunk 1 above)
Comment (michalskrivanek) [inline test/verify/check-ovirt]: looks safe enough to me
[root@ovirt ~]# rpm -q --whatrequires libxml2
libxml2-python-2.9.1-6.el7_2.3.x86_64
python-ovirt-engine-sdk4-4.1.5-2.el7.centos.x86_64
libguestfs-1.32.7-3.el7_3.3.x86_64

it's only for the ovirt image anyway

File: test/verify/check-ovirt (hunk 2 above)
Comment (michalskrivanek) [inline test/verify/check-ovirt]: I confess, I like tabs:)

File: test/verify/check-ovirt (hunk 3 above)
Comment (michalskrivanek) [inline test/verify/check-ovirt]: yes, it seems to be set up correctly and working manually, but it does require a change in the plugin as now the test machine needs to "talk" to the remote ovirt instance. Normally the deployment is always local - cockpit runs on each ovirt's host and only talks to local libvirt. 
It's a small TODO to make the libvirt URI configurable (same for regular cockpit-machines, see https://github.com/cockpit-project/cockpit/blob/83ac95f549c4ac6e18e445f41012a25e879c9bc8/pkg/machines/config.es6#L23)

File: test/verify/check-ovirt (hunk 1 above)
Comment (martinpitt) [inline test/verify/check-ovirt]: Confirmed, xmllint is in the cockpit/tests container, so should be okay.

File: bots/images/scripts/ovirt.setup (hunk 5 above)
Comment (michalskrivanek) [inline bots/images/scripts/ovirt.setup]: alrighty, I found those in other setup scripts

File: bots/images/scripts/ovirt.setup (hunk 6 above)
Comment (michalskrivanek) [inline bots/images/scripts/ovirt.setup]: ovirt is incompatible with both firewalld and networkmanager. Since part of the installation is bringing up of the system it needs to be disabled. It's normally done in the engine-setup but here I need to do that manually since the host part (vdsm) is installed in a non-standard way (otherwise it would require a dedicated machine and separate installation as part of "adding the host" - doable, but more lengthy both time-wise and code-wise)
kdump is just to speed things up as we do something with it in ovirt startup as well

File: bots/images/scripts/ovirt.setup (hunk 10 above)
Comment (michalskrivanek) [inline bots/images/scripts/ovirt.setup]: most of them are needed. The setup is typically manual. For reference I took it from https://github.com/oVirt/ovirt-containers/tree/master/image-specifications/engine which is the closest thing we have to unattended preconfigured instance.
I can clean something out knowing that we won't use those features in cockpit CI...but most of them will stay. 
Note it is a bit future-proof with the --accept-defaults option already.

File: bots/images/scripts/ovirt.setup (hunk 7 above)
Comment (michalskrivanek) [inline bots/images/scripts/ovirt.setup]: Unfortunately these are the official ovirt install repos, looking at 
http://resources.ovirt.org/pub/yum-repo/ there are only released versions and nightly build, which I wouldn't recommend to use.
A single release typically covers 3/4-1 year

File: bots/images/scripts/ovirt.setup (hunk 11 above)
Comment (stefwalter) [inline bots/images/scripts/ovirt.setup]: For better or worse, we have to do this today in our openshift image. This is a sad reality of deploying these complex products that are not subject to operating system best practices.

File: test/verify/check-ovirt (hunk 3 above)
Comment (michalskrivanek) [inline test/verify/check-ovirt]: #7484 is tracking that

File: bots/images/scripts/ovirt.setup (hunk 9 above)
Comment (michalskrivanek) [inline bots/images/scripts/ovirt.setup]: it is quite how level in fact, oVirt distributes the repo rpm which includes and enables all the required dependencies. Except for the GPG keys which need to be imported. 
I can add "--nogpgcheck" if you prefer

for reference, that ovirt-release41.rpm it currently enables following repos:
repo id                                  repo name                        status
base/7/x86_64                            CentOS-7 - Base                   9,363
centos-opstools-release/7/x86_64         CentOS-7 - OpsTools - release       424
centos-ovirt-common-testing/x86_64       CentOS-7 - oVirt common              98
centos-ovirt42-testing/x86_64            CentOS-7 - oVirt 4.2                 46
centos-qemu-ev-release/x86_64            CentOS-7 - QEMU EV                   82
extras/7/x86_64                          CentOS-7 - Extras                   449
opstools7-fluentd-012-testing/x86_64     CentOS-7 - OpsTools - fluentd te    118
opstools7-perfmon-common-testing/x86_64  CentOS-7 - OpsTools - perfmon te     66
opstools7-sensu-common-testing/x86_64    CentOS-7 - OpsTools - sensu test     77
ovirt-4.1/7                              Latest oVirt 4.1 Release          1,034
ovirt-4.1-centos-gluster38/x86_64        CentOS-7 - Gluster 3.8              226
*ovirt-4.1-epel/x86_64                   Extra Packages for Enterprise Li 11,934
ovirt-4.1-patternfly1-noarch-epel/x86_64 Copr repo for patternfly1 owned       2
ovirt-centos-ovirt41/7/x86_64            CentOS-7 - oVirt 4.1                427
ovirt-master-centos-gluster310/x86_64    CentOS-7 - Gluster 3.10              87
*ovirt-master-epel/x86_64                Extra Packages for Enterprise Li 11,934
ovirt-master-snapshot/7                  Latest oVirt master nightly snap    432
ovirt-master-snapshot-static/7           Latest oVirt master additional n    122
ovirt-master-virtio-win-latest           virtio-win builds roughly matchi     35
sac-gdeploy/x86_64                       Copr repo for gdeploy owned by s      2
updates/7/x86_64                         CentOS-7 - Updates                2,146
virtio-win-stable                        virtio-win builds roughly matchi      4

File: bots/images/scripts/ovirt.setup (hunk 8 above)
Comment (michalskrivanek) [inline bots/images/scripts/ovirt.setup]: it has it installed on filesystem, but AFAIU it still needs to be imported into RPM. Perhaps it's missing from the libguests setup?

File: bots/images/scripts/ovirt.setup (hunk 12 above)
Comment (michalskrivanek) [inline bots/images/scripts/ovirt.setup]: I'll remove this one. One of the customizations above actually bypass full host installation which can take ages since it yum installs packages on its own. Now it's just a round trip between two APIs, a fixed small delay should be fine

File: bots/images/scripts/ovirt.setup (hunk 10 above)
Comment (michalskrivanek) [inline bots/images/scripts/ovirt.setup]: actually, removed most of them, the remaining part sets up passwords and leaves out unneeded components

**Comment by michalskrivanek** [comment]
updated. fixed/hacked ovirt.setup shutdown lockup. 
Image creation and boot seems to work fine

**Comment by michalskrivanek** [comment]
updated the SSO setting to get oVirt authentication working for localhost forwards

**Comment by michalskrivanek** [comment]
@martinpitt can you please re-review? The ovirt image should be ready. The sample test can be left out and wait until @mareklibra finishes #7139. Whatever you prefer.

**Comment by petervo** [comment]
Looks there are a few things pyflakes picked up. Run 

```
tools/test-static-code
```

to see those issues.

**Comment by michalskrivanek** [comment]
ha, was missing pyflakes on my test machine

**Comment by stefwalter** [comment]
I've added a commit which builds the new image, and am uploading it.

**Comment by stefwalter** [comment]
Removed the sit() from the tests, squashed and rebased.

**Comment by stefwalter** [comment]
Test failures are not related.
